### PR TITLE
Simpler diagnostic when passing arg to closure and missing borrow

### PR DESCRIPTION
**Pull request by Akida31**
fixes #64915

I followed roughly the instructions and the older PR #76362.
The number of references for the expected and the found types will be compared and depending on which has more the diagnostic will be emitted.

I'm not quite sure if my approach with the many `span_bug!`s is good, it could lead to some ICEs. Would it be better if  those errors are ignored?

As far as I know the following code works similarly but in a different context. Is this probably reusable since it looks like it would emit better diagnostics? 
https://github.com/rust-lang/rust/blob/a688a0305fad9219505a8f2576446510601bafe8/compiler/rustc_hir_analysis/src/check/demand.rs#L713-L1061

When running the tests locally, a codegen test failed. Is there something I can/ should do about that?

If you have some improvements/ corrections please say so and I will happily include them.

r? @estebank (as you added the mentoring instructions to the issue)

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1645,6 +1647,75 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+
+            for ((found_arg, expected_arg), arg_span) in
+                found_args.zip(expected_args).zip(arg_spans)
+            {
+                let (found_ty, found_refs) = get_deref_type_and_refs(*found_arg);
+                let (expected_ty, expected_refs) = get_deref_type_and_refs(*expected_arg);
+
+                if found_ty == expected_ty {
+                    let hint = if found_refs < expected_refs {
+                        "hint: consider borrowing here:"
```
Comment (Rageking8) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1703]: Don't think there should be a need for "hint: " given that "help: " is already there.

Comment (Akida31) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: `hint` is removed in the newest commit. Any other tips? :)

**Comment by estebank** [comment]
I'll review in full soon, but I notice that the only case that is unacounted for are generators, otherwise the bug calls seem like reasonable invariant holding.

From a cursory look, I'd like it if we could move this logic to its own function to make skimming the code easier.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -1645,6 +1647,75 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+
+            for ((found_arg, expected_arg), arg_span) in
+                found_args.zip(expected_args).zip(arg_spans)
+            {
+                let (found_ty, found_refs) = get_deref_type_and_refs(*found_arg);
+                let (expected_ty, expected_refs) = get_deref_type_and_refs(*expected_arg);
+
+                if found_ty == expected_ty {
+                    let hint = if found_refs < expected_refs {
+                        "consider borrowing here:"
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1703]:
```suggestion
                        "consider borrowing the argument type"
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -1645,6 +1647,75 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+                    } else if found_refs == expected_refs {
+                        continue;
+                    } else {
+                        "consider removing the borrow:"
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1707]:
```suggestion
                        "do not borrow the argument"
```

**Comment by Akida31** [comment]
Two questions:
* There is the method [`hir::Node::fn_decl`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.Node.html#method.fn_decl), should I use that to cover generators? It would allow some other cases which are unexpected and currently `bug!`s in my implementation.
* The diagnostics of the test in `src/test/ui/closures/multiple-fn-bounds.rs` regresses with this PR, but this seems to be a constructed case which should not appear in user code. Is it fine (for now or in general) to ignore this regression?

**Comment by estebank** [comment]
> There is the method [hir::Node::fn_decl](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.Node.html#method.fn_decl), should I use that to cover generators? It would allow some other cases which are unexpected and currently bug!s in my implementation.

Yeah that 'd be better, because even if that method has a bug for generators, it'll be eventually fixed for all of its callers.

> The diagnostics of the test in src/test/ui/closures/multiple-fn-bounds.rs regresses with this PR, but this seems to be a constructed case which should not appear in user code. Is it fine (for now or in general) to ignore this regression?

That suggestion regression is acceptable.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 4]
```diff
@@ -3037,6 +3043,67 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+            };
+            err.span_suggestion_verbose(
+                arg_span,
+                hint,
+                expected_arg.to_string(),
+                Applicability::MaybeIncorrect,
+            );
+        }
+    }
+}
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:3105]: The only  thing I'd like changed is to make this a single suggestion for all arguments, but we can land as is.

**Comment by Manishearth** [comment]
@bors r- 

https://github.com/rust-lang/rust/pull/104270#issuecomment-1311263942

**Comment by estebank** [comment]
Please rebase.

**Comment by Akida31** [comment]
I added a tidy-ignore for the file-length. I don't know what the best idea would be to split this long file. Additionally I tried to combine all similar suggestions, but I'm not sure if this is the preferred way to do that with the diagnostics. Happy to make improvements :)

**Comment by estebank** [comment]
Apologies, I'd missed that you'd rebased and re-pushed. r=me after rebasing :)

**Pull request merged**